osu.Game.Rulesets.Taiko/Mods/TaikoModClassic.cs

```diff
@@ -32,7 +74,65 @@ public void Update(Playfield playfield)
             // Since the time range will depend on a positional value, it is referenced to the x480 pixel space.
             float ratio = drawableTaikoRuleset.DrawHeight / 480;
 
-            drawableTaikoRuleset.TimeRange.Value = (playfield.HitObjectContainer.DrawWidth / ratio) * scroll_rate;
+            double timeRange = (playfield.HitObjectContainer.DrawWidth / ratio) * scroll_rate;
+
+            if (mods != null)
+            {
+                if (mods.OfType<TaikoModHardRock>().Any())
+                {
+                    // For hardrock, the playfield time range is clamped to within classicMaxTimeRange and the equivalent
+                    // time range for a 16:10 aspect ratio.
+                    drawableTaikoRuleset.AdjustmentMethod.Value = AspectRatioAdjustmentMethod.None;
+                    timeRange = Math.Clamp(timeRange, DrawableTaikoRuleset.AspectRatioToTimeRange(1.6f), classic_max_time_range);
+
+                    // Scale hidden parameter to match the adjusted time range. This only affects hdhr.
+                    float hiddenRatio = 1.0f - Math.Min((float)timeRange / (float)DrawableTaikoRuleset.DEFAULT_TIME_RANGE, 1.0f);
+                    float fadeOutDuration = Math.Max(0.24f - hiddenRatio, 0.1f);
+                    hiddenInitialAlpha.Value = 0.3f * fadeOutDuration / 0.2f;
+                    hiddenFadeOutDuration.Value = fadeOutDuration;
+                }
+                else if (mods.OfType<TaikoModHidden>().Any())
+                {
+                    // Hidden aspect adjustment is overriden by hardrock in the case of hdhr, hence these are applied only
+                    // if hardrock is not enabled.
+                    drawableTaikoRuleset.AdjustmentMethod.Value = AspectRatioAdjustmentMethod.Trim;
+                    drawableTaikoRuleset.TargetAspectRatio.Value = classic_hidden_aspect_ratio;
+                }
+            }
+
+            drawableTaikoRuleset.TimeRange.Value = timeRange;
+        }
+
+        void IApplicableToDrawableHitObject.ApplyToDrawableHitObject(DrawableHitObject hitObject)
+        {
+            switch (hitObject)
+            {
+                case DrawableDrumRoll:
+                case DrawableDrumRollTick:
+                case DrawableHit:
+                    hitObject.ApplyCustomUpdateState += (o, state) =>
+                    {
+                        Debug.Assert(drawableTaikoRuleset != null);
+
+                        // For classic nomod, the effective aspect ratio will be limited to 2:1 by fading the notes in.
+                        // We do not need to explicitly detect mods here, as both hidden and hardrock limits the time
+                        // range to under the 2:1 limit.
```

Comment:
Then I assume this implicitly relies on hidden/hardrock's `ApplyCustomUpdateState` firing after `TaikoModClassic`? Bit sketchy, I don't know if anything actually enforces that. Did you check that this doesn't break if the mods are in the wrong order in the `SelectedMods` list? (i.e. `new[] { new TaikoModHidden(), new TaikoModClassic() }`)

Reply:
Nope, the time range limit is only applicable to classic hidden/hardrock, and is applied in `TaikoModClassic.Update`.